### DISCUSSION TITLE
How to ignore parts of a .dump (handling expected differences)

### DISCUSSION THREAD

**Discussion by jklingen** [pointfreeco/swift-snapshot-testing#527] [Q&A] [answered]
As a user of the SnapshotTesting library, I'd like to be able to ignore parts of a `.dump` snapshot, so that I can snapshot test a data structure containing (expectedly) varying properties. 

In my specific use case, the data structure is not entirely under my control and contains generated UUIDs. I guess other use cases might include data like generated time stamps, so I think being able to regex-replace expected differences would be a good way to go.

Is there a way to extend the existing strategy in a way that I can make modifications to the `snap`ped string before diffing? (ideally without copypasting all of [Any.swift](Sources/SnapshotTesting/Snapshotting/Any.swift) into my code base).
If yes, I'd love to learn about it. If not, I'd be happy to provide a PR enhancing Any.swift to accept an optional regex pattern.

**Comment by stephencelis** [comment, marked as answer]
@jklingen I'm going to convert this to a discussion since it's a question and not a bug report.

There are a couple ways to override a portion of a dump:

1. Under the hood, the dump strategy [uses `Mirror`](https://github.com/pointfreeco/swift-snapshot-testing/blob/9e287256b91a68dd71a0352d6947e8261193431e/Sources/SnapshotTesting/Snapshotting/Any.swift#L12), which means if you can conform the type to [`CustomReflectable`](https://developer.apple.com/documentation/swift/customreflectable) you can control the dump's hierarchy and return only the fields you want to preserve.
2. SnapshotTesting also ships with a protocol that allows you to override that portion of the dump with a string: https://github.com/pointfreeco/swift-snapshot-testing/blob/9e287256b91a68dd71a0352d6947e8261193431e/Sources/SnapshotTesting/Snapshotting/Any.swift#L66-L75

Our newer library [`swift-custom-dump`](https://github.com/pointfreeco/swift-custom-dump) offers even more features and flexibility here, which we may use in the future here. But if you need those features now, you could write your own strategy that uses CustomDump under the hood.

**Reply by jklingen** [reply]
Thanks so much for the quick and detailed response, looks very promising - I am pretty sure these options are a great fit for our use case. I will give it a try tomorrow 🙇
